### PR TITLE
Module Replication: MongoDB & MySQL support

### PR DESCRIPTION
**Pull request by rkistner**
This is a major refactor to support modular replication, and adds initial support for MongoDB and MySQL source databases.

Opening the PR to start the process of getting this into the stable builds.

### CODE REVIEW

File: modules/module-mongodb/src/replication/MongoErrorRateLimiter.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      // Wait 15 minutes, to avoid triggering Supabase's fail2ban
+      this.setDelay(900_000);
+    } else if (message.includes('ENOTFOUND')) {
+      // DNS lookup issue - incorrect URI or deleted instance
+      this.setDelay(120_000);
+    } else if (message.includes('ECONNREFUSED')) {
+      // Could be fail2ban or similar
+      this.setDelay(120_000);
+    } else if (
+      message.includes('Unable to do postgres query on ended pool') ||
```
Comment (Rentacookie) [inline modules/module-mongodb/src/replication/MongoErrorRateLimiter.ts:31]: Seems we have some lingering Postgres mentions here.

File: modules/module-mysql/package.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      "default": "./dist/types/types.js"
+    }
+  },
+  "dependencies": {
+    "@powersync/lib-services-framework": "workspace:*",
+    "@powersync/service-core": "workspace:*",
+    "@powersync/service-sync-rules": "workspace:*",
+    "@powersync/service-types": "workspace:*",
+    "@powersync/service-jsonbig": "workspace:*",
+    "@powersync/mysql-zongji": "0.0.0-dev-20241031142605",
```
Comment (Rentacookie) [inline modules/module-mysql/package.json:36]: Wondering if we should go ahead an create a Zongji release that we can use here.

File: modules/module-mysql/src/replication/MySQLErrorRateLimiter.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      // Wait 15 minutes, to avoid triggering Supabase's fail2ban
+      this.setDelay(900_000);
+    } else if (message.includes('ENOTFOUND')) {
+      // DNS lookup issue - incorrect URI or deleted instance
+      this.setDelay(120_000);
+    } else if (message.includes('ECONNREFUSED')) {
+      // Could be fail2ban or similar
+      this.setDelay(120_000);
+    } else if (
+      message.includes('Unable to do postgres query on ended pool') ||
```
Comment (Rentacookie) [inline modules/module-mysql/src/replication/MySQLErrorRateLimiter.ts:30]: This doesn't end up in the logs but the mention of postgres should be removed regardless

**Review by stevensJourney — APPROVED** [review]

File: packages/service-core/src/routes/endpoints/sync-rules.ts [hunk 4]
```diff
@@ -1,12 +1,12 @@
-import * as t from 'ts-codec';
-import type { FastifyPluginAsync } from 'fastify';
-import * as pgwire from '@powersync/service-jpgwire';
 import { errors, router, schema } from '@powersync/lib-services-framework';
 import { SqlSyncRules, SyncRulesErrors } from '@powersync/service-sync-rules';
+import type { FastifyPluginAsync } from 'fastify';
+import * as t from 'ts-codec';
 
-import * as replication from '../../replication/replication-index.js';
+import * as system from '../../system/system-index.js';
```
Comment (Rentacookie) [inline packages/service-core/src/routes/endpoints/sync-rules.ts:6]: Is this still used?

**Review by Rentacookie — APPROVED** [review]
Looks good to me. Made a few comments - mostly related to some version discrepancies with some dependencies used in the modules and service and some mentions of Postgres in the MySQL and Mongo modules.
No blockers that I could see though.

**Review by Rentacookie — APPROVED** [review]

File: modules/module-mongodb/src/replication/MongoErrorRateLimiter.ts (hunk 1 above)
Comment (stevensJourney) [reply modules/module-mongodb/src/replication/MongoErrorRateLimiter.ts]: Removed in https://github.com/powersync-ja/powersync-service/pull/137

File: modules/module-mysql/package.json (hunk 2 above)
Comment (stevensJourney) [reply modules/module-mysql/package.json]: Bumped to 0.1.0 in https://github.com/powersync-ja/powersync-service/pull/137

File: modules/module-mysql/src/replication/MySQLErrorRateLimiter.ts (hunk 3 above)
Comment (stevensJourney) [reply modules/module-mysql/src/replication/MySQLErrorRateLimiter.ts]: Removed in https://github.com/powersync-ja/powersync-service/pull/137

File: packages/service-core/src/routes/endpoints/sync-rules.ts (hunk 4 above)
Comment (stevensJourney) [reply packages/service-core/src/routes/endpoints/sync-rules.ts]: Removed in https://github.com/powersync-ja/powersync-service/pull/137

File: modules/module-postgres/package.json
[the remark this answers is not in the document]
Comment (stevensJourney) [reply modules/module-postgres/package.json:1]: Dependencies were shared and updated to common versions in https://github.com/powersync-ja/powersync-service/pull/137

**Pull request merged**